### ISSUE TITLE
Batch normalization forward with weights

### ISSUE THREAD

**Issue by adelecourot** [intel/mkl-dnn#276]
Hello, 
If i understand correctly, the weight parameter in the batch normalization forward corresponds to [gamma, beta].
I tried to use that function, but the weights do not seem to be taken into account. 
Can someone explain me why?

Here is the code I am using:

```
void batch_normalisation(int nci, int csz, int csy, int csx,
std::vector<float> * user_src, std::vector<float> * user_dst,
std::vector<float> * bn1_weights, std::vector<float> * bn1_mean,
std::vector<float> * bn1_std)
{
	auto cpu_engine = mkldnn::engine(mkldnn::engine::cpu, 0);
	std::vector<mkldnn::primitive> net;

	mkldnn::memory::dims bn_mean_tz = { nci };
	mkldnn::memory::dims bn_weights_tz = { 2*nci };
	mkldnn::memory::dims bn_tz = { 1, nci, csz, csy, csx};

	auto user_src_memory = mkldnn::memory({ { { bn_tz }, mkldnn::memory::data_type::f32,
			mkldnn::memory::format::ncdhw }, cpu_engine }, (*user_src).data());
	auto user_dst_memory = mkldnn::memory({ { { bn_tz }, mkldnn::memory::data_type::f32,
				mkldnn::memory::format::ncdhw }, cpu_engine }, (*user_dst).data());

	auto bn1_user_weights_memory = mkldnn::memory({ { { bn_weights_tz }, mkldnn::memory::data_type::f32,
					mkldnn::memory::format::x }, cpu_engine }, (*bn1_weights).data());
	auto bn1_user_mean_memory = mkldnn::memory({ { { bn_mean_tz }, mkldnn::memory::data_type::f32,
					mkldnn::memory::format::x }, cpu_engine }, (*bn1_mean).data());
	auto bn1_user_std_memory = mkldnn::memory({ { { bn_mean_tz }, mkldnn::memory::data_type::f32,
					mkldnn::memory::format::x }, cpu_engine }, (*bn1_std).data());
	auto bn_src_md = mkldnn::memory::desc(
				{ bn_tz }, mkldnn::memory::data_type::f32, mkldnn::memory::format::ncdhw);

	auto bn1_desc = mkldnn::batch_normalization_forward::desc(mkldnn::prop_kind::forward_inference,
			bn_src_md, 0.00001, 1u);
	auto bn1_pd = mkldnn::batch_normalization_forward::primitive_desc(bn1_desc, cpu_engine);

	auto bn1_weights_memory = bn1_user_weights_memory;
	auto bn1_mean_memory = bn1_user_mean_memory;
	auto bn1_std_memory = bn1_user_std_memory;
	auto bn1_src_memory = user_src_memory;
	auto bn1_dst_memory = user_dst_memory;
	std::cout << "memory" << std::endl;

	net.push_back(mkldnn::batch_normalization_forward(bn1_pd,
			(const mkldnn::primitive::at) user_src_memory,
			(const mkldnn::primitive::at) bn1_mean_memory,
			(const mkldnn::primitive::at) bn1_std_memory,
			(const mkldnn::primitive::at) bn1_weights_memory,
			(const mkldnn::memory) bn1_dst_memory));
	std::cout << "bn" << std::endl;
	mkldnn::stream(mkldnn::stream::kind::eager).submit(net).wait();
}


int main(int argc, char **argv) {
	/* Allocate input and output buffers for user data */
//	int nci = 1;
	int nco = 64;
	int csz = 32;
	int csy = 64;
	std::vector<float> bn1_weights(2*nco, 0.0f);
	std::vector<float> bn1_mean(nco, 0.0f);
	std::vector<float> bn1_std(nco, 10.0f);

	std::vector<float> user_src(nco * csz * csy * csy, 1.f);
	std::vector<float> user_dst(nco * csz * csy * csy, -32.0f);

	batch_normalisation(nco, csz, csy, csy, &user_src, &user_dst, &bn1_weights, &bn1_mean, &bn1_std);
	float min = user_dst[0];
	float max = user_dst[0];
	for(int ic=0; ic<nco; ic++){
		for (int i=0; i<csz; i++){
			for(int j=0; j<csy; j++){
				for(int h=0; h<csy; h++){
					min = min > user_dst[h + j*csy + i*csy*csy + ic*csy*csy*csz] ? user_dst[h + j*csy + i*csy*csy + ic*csy*csy*csz]:min;
					max = max < user_dst[h + j*csy + i*csy*csy + ic*csy*csy*csz]? user_dst[h + j*csy + i*csy*csy + ic*csy*csy*csz]:max;
				}
			}
		}
	}
	std::cout << min << std::endl;
	std::cout << max << std::endl;
	/* Visualization*/
	for(int ic=0; ic<2; ic++){
		for (int i=0; i<3; i++){
			for(int j=0; j<3; j++){
				for(int h=0; h<3; h++){
					std::cout<< user_dst[h + j*csy + i*csy*csy + ic*csy*csy*csz] << " ";
				}
			}
			std::cout<<std::endl;
		}
		std::cout<<std::endl;
	}
}
```

The expected results is 0, because here the scaling is 0. But what I get corresponds to gamma=1 and beta=0 (no weights in the batch normalization): (src - mean)/sqrt(std)
 here 1/sqrt(10) = 0.3162

Thank you in advance for any piece of advice.

**Comment by igorsafo** [comment]
Hello @adelecourot ,
Flag _use_scale_shift_ should be used in the parameters of batch normalization descriptor:
	_auto bn1_desc = mkldnn::batch_normalization_forward::desc(mkldnn::prop_kind::forward_inference,
			bn_src_md, 0.00001, 1u | **mkldnn::batch_normalization_flag::use_scale_shift**);_

**Comment by adelecourot** [comment]
Ah ok, I had not understood correctly the purpose of those flags. Thank you @igorsafo !
